### DISCUSSION TITLE
Console/GUI scripts with command-line arguments

### DISCUSSION THREAD

**Discussion by codeandfire** [pypa/setuptools#3352] [Q&A] [answered]
I'm not sure if this is the right place to ask this question, but what is the recommended approach to make console/GUI scripts accept command-line arguments? The current Setuptools docs on Entry Points don't mention this, but the [Python packaging user guide](https://packaging.python.org/en/latest/specifications/entry-points/#use-for-scripts) says that the functions corresponding to console/GUI scripts will be invoked with no arguments. To test this, I tried configuring a console script for the following function:
```python
def hello(foo):
    print('Hello,', foo)
```
and indeed on trying to invoke it like so:
```
$ hello 'bar'
```
the argument `foo='bar'` is not passed to the `hello()` function.

Then what is the correct way to pass command-line arguments? I checked the [Pip source code](https://github.com/pypa/pip/blob/e58a8a591b72f5f022ddd84c0c03e4bfe90d9040/src/pip/_internal/cli/main.py#L47) to see how they do it and it seems that they manipulate `sys.argv` directly:
```python
def main(args: Optional[List[str]] = None) -> int:
    if args is None:
        args = sys.argv[1:]
    ...
```
So is manipulating `sys.argv` the way to go about it?

**Comment by abravalheri** [comment, marked as answer]
Hi @codeandfire, your code should have access to `sys.argv`, that is the standard way of passing command line arguments.

The command practice however is to use a library to construct an interface for this.

If you want to stick with the standard library, you can try [`argparse`](https://docs.python.org/3/library/argparse.html).
There is a tutorial [here](https://docs.python.org/3/howto/argparse.html) and a minimal example [here](https://github.com/pyscaffold/pyscaffold-demo/blob/master/src/demo_project/skeleton.py).

It is very common to see people also using [`click`](https://click.palletsprojects.com/en/latest/) and [`typer`](https://typer.tiangolo.com/) instead of `argparse`.

**Reply by codeandfire** [reply]
@abravalheri thanks so much for your response, basically I was somehow under the impression that we are constrained to only use `sys.argv` in the console/GUI scripts setting, glad to know that we can use the regular command-line parsing utilities like argparse.

I am working on the Entry Points docs page next - do you mind if I include a line or two mentioning this point? That while the function being invoked through the console/GUI script cannot take any arguments, you can use regular command-line argument parsing utilities within the function to retrieve any input from the user. Because it might not be obvious to a newcomer - that how is user input supposed to be retrieved from within a console/GUI script.

**Reply by abravalheri** [reply]
That sounds perfect to me, thank you very much @codeandfire!
